arangod/Cluster/RebootTracker.cpp

```diff
@@ -66,296 +84,135 @@ RebootTracker::RebootTracker(RebootTracker::SchedulerPointer scheduler)
 #endif
 }
 
-void RebootTracker::updateServerState(
-    containers::FlatHashMap<ServerID, RebootId> const& state) {
-  MUTEX_LOCKER(guard, _mutex);
-
+void RebootTracker::updateServerState(State state) {
+  std::lock_guard guard{_mutex};
   LOG_TOPIC("77a6e", DEBUG, Logger::CLUSTER)
-      << "updating reboot server state from " << _rebootIds << " to " << state;
-
-  // Call cb for each iterator.
-  auto for_each_iter = [](auto begin, auto end, auto cb) {
-    auto it = begin;
-    decltype(it) next;
-    while (it != end) {
-      // save next iterator now, in case cb invalidates it.
-      next = std::next(it);
-      cb(it);
-      it = next;
-    }
-  };
-
-  // For all known servers, look whether they are changed or were removed
-  for_each_iter(_rebootIds.begin(), _rebootIds.end(), [&](auto const curIt) {
-    auto const& serverId = curIt->first;
-    auto& oldRebootId = curIt->second;
-    auto const& newIt = state.find(serverId);
-
+      << "updating reboot server state from " << _state << " to " << state;
+  for (auto it = _callbacks.begin(); it != _callbacks.end();) {
+    auto next = std::next(it);
+    auto newIt = state.find(it->first);
+    TRI_ASSERT(!it->second.empty());  // we should remove it, when it was empty
     if (newIt == state.end()) {
-      // Try to schedule all callbacks for serverId.
-      // If that didn't throw, erase the entry.
-      LOG_TOPIC("88858", INFO, Logger::CLUSTER)
```

Comment:
Maybe instead of this and "88857" messages I can write something like this
```C++
LOG_TOPIC(...) << "rebooted servers: " << std::set_difference(old_state.begin(), old_state.end(), new_state.begin(), new_state.end(), [](auto const& lhs,auto const& rhs) {
  return lhs == rhs;
});
```
If it needed of course.

I also have a feeling that we need log something like this in cluster info, not in reboot tracker

Reply:
So regarding the log messages. I'd like to see the following:

* On INFO: When another server changes (reboots or is removed), a message when that is noticed, and which server specifically this was about. This was previously handled by `88857` and `88858`.
  This is often useful when analyzing log files.
* On DEBUG: When callbacks are executed, and which ones. This is handled by `afdfd` as before, but is now set to TRACE. I think it'd be better on DEBUG, in particular with the next point:
  This is useful when debugging something particular, e.g. why some cleanup doesn't happen or why some server is not releasing its locks or so.
* On TRACE: Changes to state as logged by `77a6e`. This was previously DEBUG (not changed by this PR), but I think it's the least useful to users, but the most noisy one, so should be changed to TRACE.

WDYT?

Reply:
So I see 3 questions:
1. Do we really need this log messages, or maybe only rebooted servers that had callbacks is enough?
2. What and why should be log level for log old and new state: info, debug, trace?
3. Is difference between iteration for `_callbacks` or `_state` important? 
    `_callbacks` size is `O(1)`, it's leaders and some other important servers.
    `_state` size is count of servers in cluster.
4. if 1 and 3 yes, should we think about log level higher than info? Or readability and simplicity is preferable?

Reply:
Done?

Reply:
> Do we really need this log messages, or maybe only rebooted servers that had callbacks is enough?

Maybe. What I can say is: These log messages (88857 and 88858) have been extremely helpful in the past to debug different problems, both in support and development.

If it's always there, it helps to align (distributed) logs. That's not possible when the message isn't there.

Another drawback with logging only those will callbacks is that an expected log message doesn't appear, and you don't know whether it's because the server didn't notice the reboot, or just didn't have any callbacks.

How often that makes an impact I cannot say.

So my problem here is: I know that the logs are helpful the way they are. The will surely lose some of that, and maybe a lot: I don't know. Either way, I don't think it's worth losing them, because there's no visible improvement on the other side.

> What and why should be log level for log old and new state: info, debug, trace?

I don't get the question, I thought I've written that in the previous post? Can you be more specific?

> Is difference between iteration for _callbacks or _state important?
> _callbacks size is O(1), it's leaders and some other important servers.
> _state size is count of servers in cluster.

Regarding __callbacks size is O(1)_: I don't see how you arrive at that conclusion, unless you mean O(1) because it'll at most be 30 or so. Perhaps clarify what you mean by "leader"? That term's pretty overloaded. Note that if there are a few AQL queries running, that alone will usually add all the Coordinators on all DBServers; and all DBServers on all Coordinators.

Anyway, to answer the question: There will be a difference in size. But the amount of servers is so small that it does not matter. Also note that this is usually executed at most once a second by the heartbeat thread; and its complexity dwarfs in comparison to all the other stuff done by `ClusterInfo::flush()` (where this is called).

> if 1 and 3 yes, should we think about log level higher than info? Or readability and simplicity is preferable here?

When you say higher, do you mean WARN etc.? I can't think of anything that should be a warning here. And I don't understand the second question in this context?

Reply:
> When you say higher, do you mean WARN etc.? I can't think of anything that should be a warning here. And I don't understand the second question in this context?

This question was about users that have more higher log level than INFO, so them don't have INFO in logs
But it's outdated

Reply:
> I don't get the question, I thought I've written that in the previous post? Can you be more specific?

Yes, you already answered, I didn't see it, I also wrote about that